components/tikv_util/src/macros.rs

```diff
@@ -171,6 +171,29 @@ macro_rules! try_opt_or {
     }};
 }
 
+/// It checks whether the current thread is unwinding because of panic. If it is,
+/// log an error message instead of causing double panic.
+#[macro_export]
+macro_rules! safe_panic {
+    () => ({
+        safe_panic!("explicit panic")
+    });
+    ($msg:expr) => ({
+    if std::thread::panicking() {
+            error!(concat!($msg, " at {}:{}, double panic prevented"), file!(), line!())
+        } else {
+            panic!($msg);
+        }
+    });
+    ($fmt:expr, $($args:tt)+) => ({
+        if std::thread::panicking() {
+            error!(concat!($fmt, " at {}:{}, double panic prevented"), $($args)+, file!(), line!())
```

Comment:
How about using slog style? `"file" => file!(), "line" => line!()`

Reply:
I just realized slog prints file and line to the log, so I remove it. Please take a look again.